### ISSUE TITLE
Canvas drawstate

### ISSUE THREAD

**Issue by emilwestergren** [magic-lang/ooc-kean#857]
The number of available draw parameters when using a `Canvas` has constantly increased. To simply overload functions doesn't scale well with this. It is clear that we need some kind of state which the user can tweak before the actual draw call. We *kind of* have this already since `Canvas` does have a few states that can be changed. The problem here is that since we reuse the same `Canvas` in multiple filters, we only want the state to be temporary. Otherwise we get strange bugs and ugly code where we need to manually reset the state to default before returning:
```ooc
image canvas focalLength = focalLength
image canvas draw(input)
image canvas focalLength = 0.0f
```

Me and Simon had a discussion about this a while ago and I believe two suggestions came up.

**DrawState**
The parameters are gathered in a cover which is used as a parameter to the Draw API. 
```ooc
DrawState: cover {
	focalLength := 0.0f
	transform := FloatTransform3D identity
	opacity := 1.0f
	source, destination: IntBox2D
	init: func@ (=source, =destination)
	init: func@ ~IntSize2D (source, destination: IntSize2D) { this init(IntBox2D new(source), IntBox2D new(destination)) }
	init: func@ ~fromImage (source: Image, destination: Canvas) { this init(IntBox2D new(source size), IntBox2D new(destination size))}
}

state := image canvas defaultState
state focalLength = 1500.0f
state opacity = 0.5f
image canvas draw(input, state)
```

**SubCanvas**
This is similiar to `DrawState` but the state stays in the `Canvas`. Instead of creating a `DrawState`, a `SubCanvas` is created with default state. The state can then be altered on this subcanvas before it's used to draw. One problem here could be that we probably want the `SubCanvas` to be a cover so we don't have to deal with freeing it every time.

```ooc
subcanvas := image canvas createSubCanvas() //or image canvas always returns a subcanvas
subcanvas focalLength = 1500.0f
subcanvas opacity = 0.5f
subcanvas draw(input)
```
I know @simonmika wants to go for **SubCanvas** and I think it could work. @sebastianbaginski @davidpiuva any opinions?

Of course anyone else is also welcome to express their opinion.

**Comment by sebastianbaginski** [comment]
I think subcanvas fits current design better and this is what Simon suggested I should do with `Raster` classes. I'm fine with both solutions though.

**Comment by davidpiuva** [comment]
I prefer DrawState because I can think of it as a brush instead of another canvas.

**Comment by emilwestergren** [comment]
If we go for **SubCanvas**: Do we just deal with it being a class and free it after use?
```ooc
subcanvas := image createCanvas() 
subcanvas draw(input)
subcanvas free()
```

It would be nice to have a cover solution and not have to free it, then again the memory size of `Canvas` / `DrawState` is probably a bit too large for a cover.

**Comment by davidpiuva** [comment]
It depends on how you use the cover. Passed by reference, it is just like a class.

**Comment by emilwestergren** [comment]
Of course, but are we "allowed" to do that?

**Comment by davidpiuva** [comment]
The only problem is the low level syntax in ooc that can't take references automatically.

**Comment by emilwestergren** [comment]
I know, it would look something like this:
```ooc
draw: func (state: DrawState@)
...
image canvas draw(myState&)
```

**Comment by davidpiuva** [comment]
Even C++ is cleaner in handling references.

**Comment by simonmika** [comment]
A subcanvas is a inherits `canvas` or at least `surface`. It is going to be a class. Can the one who does this run the API by me, please?

**Comment by davidpiuva** [comment]
The final PR (https://github.com/magic-lang/ooc-kean/pull/1226) became a mix of both solutions. It is a cover that can be passed as an argument so that no garbage collection is needed but the DrawState also contains a draw function to get the same usability as a SubCanvas would have.

**Comment by marcusnaslund** [comment]
Is this fixed now?

**Comment by davidpiuva** [comment]
DrawState can be used but is missing some minor features preventing a full migration to it.

**Comment by davidpiuva** [comment]
Half of the codebase now uses DrawState instead of canvas.

**Comment by marcusnaslund** [comment]
What's left to do here now? Can you make a short list?

**Comment by davidpiuva** [comment]
Some old bugs from before DrawState were found with focalLength so the list became longer.

Implement blend in a better way.

Put focalLength inside of DrawState working with both reference and normalized coordinates.
Fix the special cases where focalLength is off between Y and UV.
Adapt the final code to drawstate.

**Comment by davidpiuva** [comment]
The remaining bug is seen when rendering with a projective transform on Y and UV at the same time. The UV channel is currently normalized using the Y channel's resolution but later un-normalized using its own UV resolution. The problem is that the system was meant to work with normalised coordinates so I am making another attempt to normalize all the matrix multiplications with focalLength included in DrawState.

**Comment by davidpiuva** [comment]
The last bug seems to be fixed by removing a bug from transform conversions in https://github.com/magic-lang/ooc-kean/pull/1444.

**Comment by davidpiuva** [comment]
The PR https://github.com/magic-lang/ooc-kean/pull/1453 will remove most of the legacy draw calls. Only line drawing, point drawing and CPU drawing remains of the old API.
Later I will try to find unused variables in Canvas and see if something require another DrawState adaptation to be deleted.

**Comment by marcusnaslund** [comment]
Since it's been a month, please update with the status after #1639 @davidpiuva

**Comment by davidpiuva** [comment]
The remaining thing is to implement DrawState for the CPU that has been waiting for a long time from low priority.
